torch/optim/optimizer.py

```diff
@@ -293,7 +284,7 @@ def add_param_group(self, param_group):
             if not isinstance(param, torch.Tensor):
                 raise TypeError("optimizer can only optimize Tensors, "
                                 "but one of the params is " + torch.typename(param))
-            if not param.is_leaf:
+            if not self.defaults.get('differentiable', False) and not param.is_leaf:
```

Comment:
Do we want to remove the `is_leaf` restriction entirely or is there still some reason to keep it around? cc @albanD

Reply:
The original reason is still there: you should not give a Tensor to your optimizer that will not get a .grad field.
I think a good way to relax this would be to do `not (param.is_leaf or param.retains_grad)` for the second part of the condition. And disabling it for differentiable sounds good for sure.